pkg/build/starnix.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package build
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/google/syzkaller/pkg/osutil"
+	"github.com/google/syzkaller/sys/targets"
+	starnixVM "github.com/google/syzkaller/vm/starnix"
+)
+
+type starnix struct{}
+
+// The name of the Fuchsia assembly override defined in `overrideRule`.
+const overrideName = "syzkaller_starnix"
+
+// A Fuchsia assembly override definition adding a package containing fuzzing dependencies.
+var overrideRule = fmt.Sprintf(`import("//build/assembly/developer_overrides.gni")
+
+assembly_developer_overrides("%s") {
+    testonly = true
+    base_packages = [
+        "//src/testing/fuzzing/syzkaller/starnix:syzkaller_starnix",
+    ]
+}
+`, overrideName)
+
+func (st starnix) build(params Params) (ImageDetails, error) {
+	sysTarget := targets.Get(targets.Linux, params.TargetArch)
+	arch := sysTarget.KernelArch
+	if arch != "x86_64" {
+		return ImageDetails{}, fmt.Errorf("unsupported starnix arch %v", arch)
+	}
+	arch = "x64"
+	product := fmt.Sprintf("%s.%s", "workbench_eng", arch)
+
+	localDir := filepath.Join(params.KernelDir, "local")
+	if err := os.MkdirAll(filepath.Join(params.KernelDir, "local"), 0755); err != nil {
+		return ImageDetails{}, err
+	}
+	overridePath := filepath.Join(localDir, "BUILD.gn")
+	if err := os.WriteFile(overridePath, []byte(overrideRule), 0660); err != nil {
+		return ImageDetails{}, err
+	}
+	if err := osutil.SandboxChown(overridePath); err != nil {
+		return ImageDetails{}, err
+	}
+	if err := osutil.SandboxChown(localDir); err != nil {
+		return ImageDetails{}, err
+	}
+	if _, err := runSandboxed(
+		time.Hour,
+		params.KernelDir,
+		"scripts/fx", "--dir", "out/"+arch,
+		"set", product,
+		"--assembly-override", fmt.Sprintf("//products/workbench/*=//local:%s", overrideName),
+	); err != nil {
+		return ImageDetails{}, err
+	}
+
+	imageFilePath := filepath.Join(params.OutputDir, "image")
+	imageFile, err := os.Create(imageFilePath)
```

Comment:
Why do we need to create an empty file here?

Reply:
The short answer is that syz-ci's image test checks that this file exists, and the image test fails if not. The contents of the file are not used by the test, so an empty file here is enough to get syz-ci working. 

Sorry that I didn't look for a better solution here yet. I think there might be an unintended interaction between two methods involved in the image test, leading to an unnecessary(?) check for this file.

The sequence is:

- syz-ci's `Manager.build()` indirectly calls `Manager.createTestConfig()` with `imageDir` = `tmpDir` (which corresponds to the `outputDir` in this build script at the time of the initial build).
- `Manager.createTestConfig()` partially populates a `mgrconfig.Config`. (It does not populate the Image field.) Then, `Manager.createTestConfig()` calls `instance.SetConfigImage()` with the partially-populated struct and the `imageDir` string.
- `instance.SetConfigImage()` sets the `mgrconfig.Config`'s `Image` field to "${tmpDir}/image", so that the `Image` field is no longer empty. The slightly-more-populated config struct is returned back to `createTestConfig()`.
- Later, `createTestConfig()` calls `mgrconfig.Complete()` on the partially-populated config struct. That method does the following check:
```
if cfg.Image != "" {
		if !osutil.IsExist(cfg.Image) {
			return fmt.Errorf("bad config param image: can't find %v", cfg.Image)
		}
		cfg.Image = osutil.Abs(cfg.Image)
}
```
Since the config's `Image` field is not empty if it has come from `createTestConfig()`, we hit the existence check in that case.

My guess is that the right solution is to create the image path in `instance.SetConfigImage()` and check whether that file exists, and proceed without setting the config's `Image` field if not.

Reply:
Sorry for the essay / misunderstanding! No longer creating an empty file here.

Reply:
No worries! Just to conclude this, we discovered that the empty image had been necessary because of a bug preventing the actual image from being copied. That bug is now fixed.